staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter.go

```diff
@@ -389,18 +390,24 @@ func interfaceFromUnstructured(sv, dv reflect.Value) error {
 }
 
 func (c *converterImpl) ToUnstructured(obj interface{}) (map[string]interface{}, error) {
-	t := reflect.TypeOf(obj)
-	value := reflect.ValueOf(obj)
-	if t.Kind() != reflect.Ptr || value.IsNil() {
-		return nil, fmt.Errorf("ToUnstructured requires a non-nil pointer to an object, got %v", t)
+	var u map[string]interface{}
+	var err error
+	if unstr, ok := obj.(runtime.Unstructured); ok {
+		u = DeepCopyJSON(unstr.UnstructuredContent()).(map[string]interface{})
+	} else {
+		t := reflect.TypeOf(obj)
+		value := reflect.ValueOf(obj)
+		if t.Kind() != reflect.Ptr || value.IsNil() {
+			return nil, fmt.Errorf("ToUnstructured requires a non-nil pointer to an object, got %v", t)
+		}
+		u = map[string]interface{}{}
+		err = toUnstructured(value.Elem(), reflect.ValueOf(&u).Elem())
 	}
-	u := &map[string]interface{}{}
-	err := toUnstructured(value.Elem(), reflect.ValueOf(u).Elem())
 	if c.mismatchDetection {
-		newUnstr := &map[string]interface{}{}
-		newErr := toUnstructuredViaJSON(obj, newUnstr)
+		newUnstr := map[string]interface{}{}
+		newErr := toUnstructuredViaJSON(obj, &newUnstr)
 		if (err != nil) != (newErr != nil) {
-			glog.Fatalf("ToUnstructured unexpected error for %v: error: %v", obj, err)
+			glog.Fatalf("ToUnstructured unexpected error for %v: error: %v; newErr: %v", obj, err, newErr)
```

Comment:
why not return the error?

Reply:
It is an assertion, so IMHO a panic is the most appropriate thing to do. See #49369.
I can change it to return this error, I don't mind.

Reply:
This whole if-clause looks suspicious. I guess it is there to support other types which happen to create the same JSON output. I am fine with leaving it is for now. It would be worth a follow-up to try to understand this code better.